sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer.py

```diff
@@ -0,0 +1,221 @@
+                self.flush(timeout=timeout)
+
+            if timeout_time and time.time() > timeout_time:
+                raise OperationTimeoutError("Failed to enqueue events into buffer due to timeout.")
+
+            try:
+                # add single event into current batch
+                self._cur_batch.add(events)
+            except AttributeError:
+                # if the input events is a EventDataBatch, put the whole into the buffer
+                self._buffered_queue.put(events)
+                # create a new batch for incoming events
+                self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+                # put the new batch into the buffer
+                self._buffered_queue.put(self._cur_batch)
+            except ValueError:
+                # add single event exceeds the cur batch size, create new batch
+                self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+                self._buffered_queue.put(self._cur_batch)
+                self._cur_batch.add(events)
+            self._cur_buffered_len += new_events_len
+            # notify the max_wait_time worker
+            self._not_empty.notify()
+
+    def enhanced_callback_decorator(self, callback):
+        def wrapper_callback(*args, **kwargs):
+            try:
+                callback(*args, **kwargs)
+            except Exception as exc:
+                _LOGGER.warning("callback exception", callback.__name__, exc, self.partition_id)
+
+        return wrapper_callback
+
+    def sent_callback(self, events, future: Future):
+        try:
+            future.result()
+            self._on_success(events, self.partition_id)
+        except Exception as exc:
+            self._on_error(events, exc, self.partition_id)
+        finally:
+            self._cur_buffered_len -= len(events)
+            self._max_concurrent_sends_semaphore.release()
+
+    def flush(self, timeout=None, raise_error=True):
+        # try flushing all the buffered batch within given time
+        _LOGGER.info("Partition: {} started flushing.".format(self.partition_id))
+        with self._not_empty:
+            timeout_time = time.time() + timeout if timeout else None
+            while not self._buffered_queue.empty():
+                remaining_time = timeout_time - time.time() if timeout_time else None
+                # If flush could get the semaphore, perform sending
+                if ((remaining_time and remaining_time > 0) or remaining_time is None) and \
+                        self._max_concurrent_sends_semaphore.acquire(timeout=remaining_time):
+                    batch = self._buffered_queue.get()
+                    if not batch:
+                        self._max_concurrent_sends_semaphore.release()
+                        continue
+                    try:
+                        _LOGGER.info("Partition {} is sending.".format(self.partition_id))
+                        self._producer.send(
```

Comment:
Shouldn't this be a future added to the executor, rather than sent synchronously?
Otherwise where does the 'max_concurrency' take effect?

Reply:
yes, that was originally I thought of to implement the feature. this will be addressed in the issue: https://github.com/Azure/azure-sdk-for-python/issues/23844